### PR TITLE
web: add CSS-modules support

### LINKED ISSUE #19953: Add CSS modules support
Configure Webpack loaders for CSS modules based on #19349.

### PR DESCRIPTION
**Pull request by valerybugakov**
## Changes

- Added configuration for CSS modules as per [RFC](https://github.com/sourcegraph/sourcegraph/issues/19349).
- Updated Webpack configs for a web app, browser extension, and Storybook.
- Changed `BrandedStory` CSS injection: instead of inlining style tag after story content, it now prepends style tag to the document head. It's done to keep CSS modules priority higher compared to other styles.
- Refactored `PageSelector` component to CSS modules.

## Note

For some reason, Chromatic shows diffs for multiple stories. Probably it's related to `BrandedStory` changes. But after looking into these stories with dev tools, I cannot find any differences with stories from main. Let me know if you see how they are different.

Closes #19953.

### CODE REVIEW

**Comment by limitedmage** [comment]
> For some reason, Chromatic shows diffs for multiple stories. Probably it's related to `BrandedStory` changes. But after looking into these stories with dev tools, I cannot find any differences with stories from main. Let me know if you see how they are different.

Looks like differences in padding/margin?

**Comment by valerybugakov** [comment]
@limitedmage right, but they are visible only on screenshots for me. If I open two versions of the same story and check padding/margin with dev tools, they are identical 🧐

File: client/wildcard/src/components/PageSelector/PageSelector.tsx [hunk 1]
```diff
@@ -5,9 +5,11 @@ import ChevronRightIcon from 'mdi-react/ChevronRightIcon'
...
 import useResizeObserver from 'use-resize-observer'
 
-import { createAggregateError } from '../../../../shared/src/util/errors'
-import { useOffsetPagination } from '../../hooks'
-import { useDebounce } from '../../hooks/useDebounce'
+import { createAggregateError } from '@sourcegraph/shared/src/util/errors'
+
+import { useOffsetPagination, useDebounce } from '../../hooks'
+
+import styles from './PageSelector.module.scss'
```
Comment (eseliger) [inline client/wildcard/src/components/PageSelector/PageSelector.tsx:12]: I wish this was part of the LSIF dumps.. 🤔 @sourcegraph/code-intel I wonder if it can pick up the d.ts file that's generated as part of the build process

File: client/wildcard/src/components/PageSelector/PageSelector.module.scss.d.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// This file is automatically generated.
```
Comment (eseliger) [inline client/wildcard/src/components/PageSelector/PageSelector.module.scss.d.ts:1]: do these files need to end up in git? The changes to these files will most likely just be noise

**Review by felixfbecker — COMMENTED** [review]
It would be good to add some documentation to our [build process](https://docs.sourcegraph.com/dev/background-information/web/build) docs and maybe in some comments in the config documenting how this works, what the conventions are and when to use what. For example, seems like files named `.module.scss` will need to be imported into TSX, while plain `.scss` files need to be imported for "global" styles or into other `.scss` files for variables.

File: client/wildcard/src/components/PageSelector/PageSelector.tsx (hunk 1 above)
Comment (felixfbecker) [reply client/wildcard/src/components/PageSelector/PageSelector.tsx:12]: It would be if it's generated as part of `gulp generate`, which we run in the GitHub actions before running the LSIF indexer. But I don't see an edit to the Gulpfile, so not sure how this is currently generated 🤔

Comment (felixfbecker) [reply client/wildcard/src/components/PageSelector/PageSelector.tsx:12]: It would be possible to add more commands there too, although a full webpack build would probably increase the build time by a lot...

File: client/web/webpack.config.js [hunk 3]
```diff
@@ -156,24 +173,34 @@ const config = {
...
             },
           },
-        ],
+          {
+            loader: 'css-loader',
+            options: {
+              sourceMap: isDevelopment,
+              localsConvention: 'camelCase',
+              modules: {
+                localIdentName: isDevelopment ? '[name]__[local]' : '[hash:base64]',
```
[the remark this answers is not in the document]
Comment (valerybugakov) [reply client/web/webpack.config.js:194]: After some research it seems that `'[name]__[local]_[hash:base64:5]'` for both environments in a good compromise.

File: client/web/webpack.config.js [hunk 4]
```diff
@@ -156,24 +173,34 @@ const config = {
...
         test: /\.(sass|scss)$/,
-        use: [
-          mode === 'production' ? MiniCssExtractPlugin.loader : 'style-loader',
-          'css-loader',
-          'postcss-loader',
+        include: /\.module\.(sass|scss)$/,
+        use: getCSSLoaders(
           {
-            loader: 'sass-loader',
+            loader: 'css-modules-typescript-loader',
```
[the remark this answers is not in the document]
Comment (valerybugakov) [reply client/web/webpack.config.js:183]: This loader's main purpose is to create TS declaration files for CSS modules so they can be safely used in our Typescript components. It's not involved in type checking. Check out more details [here](https://github.com/seek-oss/css-modules-typescript-loader).

File: client/wildcard/src/components/PageSelector/PageSelector.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (valerybugakov) [reply client/wildcard/src/components/PageSelector/PageSelector.tsx:12]: For `Props` it's useful because often we want to reuse `Props` of one component in the type declaration of another one. Then having `PageSelectorProps` and `ButtonProps` helps with readability. It should not apply to CSS modules. If we want to import `styles` from another component, we probably need to refactor our components to avoid such a dependency.

I would personally avoid it for `styles` because then it will be too verbose in most cases: e.g., `pageSelectorClassNames.pageSelectorButton`. 

Curious to hear other POVs. 
cc @sourcegraph/frontend-platform

File: client/wildcard/src/components/PageSelector/PageSelector.module.scss.d.ts (hunk 2 above)
Comment (valerybugakov) [reply client/wildcard/src/components/PageSelector/PageSelector.module.scss.d.ts:1]: These files are required to run `tsc` without a dependency on the Webpack build command because Webpack generates these type-declarations. Without them `tsc` will fail on CI because of the untyped `styles` object. 

More information about it [here](https://github.com/seek-oss/css-modules-typescript-loader#verify-mode).

Other generated files are handled by `gulp generate` command. I looked into it briefly and didn't find a way to quickly adopt the same approach.

File: client/wildcard/src/components/PageSelector/PageSelector.tsx (hunk 1 above)
Comment (valerybugakov) [reply client/wildcard/src/components/PageSelector/PageSelector.tsx:12]: [Reply](https://github.com/sourcegraph/sourcegraph/pull/19919#discussion_r612309558) about how type-declarations are generated. 
Can you elaborate on `LSIF dumps`, please? I didn't dig into this part of functionality yet.

File: client/wildcard/src/components/PageSelector/PageSelector.module.scss.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (valerybugakov) [reply client/wildcard/src/components/PageSelector/PageSelector.module.scss.d.ts:1]: Let's continue the discussion in [this thread](https://github.com/sourcegraph/sourcegraph/pull/19919#discussion_r612309558).

File: client/wildcard/src/components/PageSelector/PageSelector.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (felixfbecker) [reply client/wildcard/src/components/PageSelector/PageSelector.tsx:12]: It is also useful when there are multiple components in one file (even if you add one later, you don't need to rename everything). Especially for props where interfaces get merged by TypeScript silently.
But also wanna hear other people's thoughts.

File: client/wildcard/src/components/PageSelector/PageSelector.module.scss.d.ts (hunk 2 above)
Comment (felixfbecker) [reply client/wildcard/src/components/PageSelector/PageSelector.module.scss.d.ts:1]: The README mentions that "there are many different solutions for this problem". I wonder if there is a solution (typings generator) that would generate typings for the CSS modules outside of webpack, that we can integrate into the other codegen steps and so we don't have to run a full Webpack build for typechecking/LSIF generating (but without checking in the typings into source control).

If you already investigated some, mind sharing what you tried? If there is no easy alternative webpack seems fine of course, it would just add analysis time and differs from how our codegen usually is done.

Comment (valerybugakov) [reply client/wildcard/src/components/PageSelector/PageSelector.module.scss.d.ts:1]: Interesting, I didn't dig a lot in the direction of standalone generators. Quick googling shows some good results: https://github.com/skovy/typed-scss-modules. I'll look into it, thanks for the suggestion.

Comment (valerybugakov) [reply client/wildcard/src/components/PageSelector/PageSelector.module.scss.d.ts]: @felixfbecker @eseliger take a look at the new approach with a standalone generator for CSS modules type declarations.

File: client/wildcard/src/components/PageSelector/PageSelector.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (pdubroy) [reply client/wildcard/src/components/PageSelector/PageSelector.tsx:12]: The convention I have seen is to use `styles`. I don't see the need to use something more verbose in the common case — local renaming is pretty trivial.

File: client/shared/dev/generateCssModulesTypes.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
+// @ts-check
+
+const { spawn } = require('child_process')
+const path = require('path')
+
+const CSS_MODULES_GLOB = path.resolve(__dirname, '../../*/src/**/*.module.scss')
+const TSM_LOG_LEVEL = '--logLevel error'
+const TSM_COMMAND = 'yarn --silent --ignore-engines --ignore-scripts tsm'
```
Comment (felixfbecker) [inline client/shared/dev/generateCssModulesTypes.js:8]: Don't these arguments have to be in the `args` array? Afaik the first argument is (only) the binary

File: client/wildcard/src/components/PageSelector/PageSelector.tsx (hunk 1 above)
Comment (eseliger) [reply client/wildcard/src/components/PageSelector/PageSelector.tsx:12]: We have a [github action that uploads an `LSIF` dump](https://github.com/sourcegraph/sourcegraph/blob/main/.github/workflows/lsif.yml) of our code to Sourcegraph.com, which enables precise code intel in PR reviews 
It seems like it still doesn't have any hover information for these fields, not sure why 🤔

**Review by eseliger — APPROVED** [review]
LGTM, what's the tl;dr on the migration plan for existing components? Are the owning teams expected to do that and by when do we want to have it fully migrated (and potentially forbid direct scss imports)? 
Also @felixfbecker maybe a code insight can track this (nerdsnipe?) 🤓 

If we could get LSIF to work before the merge, that'd be dope 👌 I would like sg/sg to be a great example repo for how nice precise code intel can be and how far it can go :) (Maaaybe, that could also mean that the css declaration content is also in a multi line comment attached to the property, so you can see the underlying css in the hover tooltip? 😉 That is definitely not required for this PR though, but maybe _someone_ finds it interesting and wants to work on this as a hack project 😬 )

File: client/wildcard/src/components/PageSelector/PageSelector.module.scss.d.ts (hunk 2 above)
Comment (felixfbecker) [reply client/wildcard/src/components/PageSelector/PageSelector.module.scss.d.ts]: Nice!

File: client/wildcard/src/components/PageSelector/PageSelector.tsx (hunk 1 above)
Comment (felixfbecker) [reply client/wildcard/src/components/PageSelector/PageSelector.tsx:12]: Maybe @sourcegraph/code-intel has some insight?

Comment (valerybugakov) [reply client/wildcard/src/components/PageSelector/PageSelector.tsx:12]: Guess: might it be that LSIF doesn't work for `styles` because we don't have `lsif-tsc-wildcard` job [here](https://github.com/sourcegraph/sourcegraph/blob/main/.github/workflows/lsif.yml). Wildcard package contains the only CSS module that we have and type declarations for it.

**Comment by pdubroy** [comment]
@eseliger:

> If we could get LSIF to work before the merge, that'd be dope 👌 I would like sg/sg to be a great example repo for how nice precise code intel can be and how far it can go :)

I totally agree that it would be cool to get precise code intel working, but I do want to be careful about feature creep in PRs. As our code review guidelines say "[Don't] Submit PRs with multiple overlapping concerns" and "PRs tackling multiple issues at once should be split into more highly-focused PRs when possible." So unless this is a regression (I don't think it is?) or a trivial addition (5-10 mins of work and a few lines of code), then I think it's best done in a separate PR.

File: client/web/webpack.config.js [hunk 6]
```diff
@@ -37,12 +39,34 @@ const babelLoader = {
 
 const extensionHostWorker = /main\.worker\.ts$/
 
+/**
+ * Generates array of CSS loaders both for regular CSS and CSS modules.
+ * Useful to ensure that we use the same configuration for shared loaders: postcss-loader, sas-loader, etc.
```
Comment (umpox) [inline client/web/webpack.config.js:44]:
```suggestion
 * Useful to ensure that we use the same configuration for shared loaders: postcss-loader, sass-loader, etc.
```

:)

File: client/web/webpack.config.js [hunk 7]
```diff
@@ -37,12 +39,34 @@ const babelLoader = {
...
+const getCSSLoaders = (...loaders) => [
+  isDevelopment ? 'style-loader' : MiniCssExtractPlugin.loader,
+  ...loaders,
+  'postcss-loader',
+  {
+    loader: 'sass-loader',
+    options: {
+      sassOptions: {
+        implementation: require('sass'),
+        includePaths: [nodeModulesPath],
```
Comment (umpox) [inline client/web/webpack.config.js:58]: Aside from this PR, do we know why we include `node_modules` here? Seems a bit odd to me, and bad for performance

File: client/wildcard/src/components/PageSelector/PageSelector.tsx [hunk 8]
```diff
@@ -16,7 +18,7 @@ interface PageButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement>
 const PageButton: React.FunctionComponent<PageButtonProps> = ({ children, active, ...props }) => (
     <button
         type="button"
-        className={classNames('btn', active ? 'btn-primary' : 'btn-link', 'mx-1', 'page-selector__button')}
+        className={classNames('btn', active ? 'btn-primary' : 'btn-link', 'mx-1', styles.pageSelectorButton)}
```
Comment (umpox) [inline client/wildcard/src/components/PageSelector/PageSelector.tsx:21]: Might be worth refactoring these styles away from BEM? As this is already scoped, this could just be `styles.button`

**Review by umpox — APPROVED** [review]
Awesome

**Comment by valerybugakov** [comment]
I would definitely want to enable LSIF support for CSS modules in our repo, but from my reading about LSIF infrastructure, we need to update Github's LSIF actions to include the `wildcard` package. I will chat today with @eseliger about that. I agree with @pdubroy that it's best to be done in a separate PR. 

### TLDR on the migration plan

> Are the owning teams expected to do that

Owning teams are expected to use CSS modules for new components and refactorings if it's doesn't critically affect deadlines. After merging this PR, I will post an announcement about that to #dev-chat to make everybody aware of this approach if they missed the PR.

> by when do we want to have it fully migrated

I'm not sure about the deadline for the migration, but it would be great to have a tool to monitor our migration process and linter rules that will encourage CSS modules' usage over global CSS rules.

**Comment by felixfbecker** [comment]
@pdubroy I think there may be a misunderstanding here. We **already** have precise code intel for all of our TypeScript code, including for generated code, and most of us rely on that heavily when reviewing PRs (through our browser extension). So in a way that is an expectation or "standard" for all code added (which is usually fulfilled automatically, unless it's some new kind of generated code, a new project, etc). This PR adds generated TypeScript code, but codeintel doesn't seem to be working on it, so the suggestion here was only "can we investigate why it doesn't work on this newly added, generated code before merging?", not "can we set up LSIF from scratch for all of our code", which I agree would definitely be orthogonal to this PR.

> I would definitely want to enable LSIF support for CSS modules in our repo, but from my reading about LSIF infrastructure, we need to update Github's LSIF actions to include the wildcard package. I will chat today with @eseliger about that. I agree with @pdubroy that it's best to be done in a separate PR.

Ah, that explains it. So this was just an oversight when adding the `wildcard` folder. In that case it seems totally fine to add that in a separate PR, as it had nothing to do with this PR and I would assume the codeintel for the generated types already work in any other folder that is _not_ the `wildcard` folder (the example just happened to be in `wildcard`).

You should be able to just copy-paste the action for the other folders so this should be easy to fix (in another PR).

**Comment by felixfbecker** [comment]
> > Are the owning teams expected to do that
> 
> Owning teams are expected to use CSS modules for new components and refactorings if it's doesn't critically affect deadlines. After merging this PR, I will post an announcement about that to #dev-chat to make everybody aware of this approach if they missed the PR.

If that is the expectation, could you state that in the docs? Currently it's worded like "you can chose to use this approach if you want to, but there is no expectation". Slack messages are ephemeral, the docs are the source of truth :)
Would also recommend #dev-announce and/or #dev-frontend as #dev-chat is very noisy.

> > by when do we want to have it fully migrated
> 
> I'm not sure about the deadline for the migration, but it would be great to have a tool to monitor our migration process and linter rules that will encourage CSS modules' usage over global CSS rules.

I can help you set up a code insight for that! Great dogfood opportunity and insights can easily do exactly this :)

File: client/storybook/src/main.ts [hunk 9]
```diff
@@ -106,23 +120,25 @@ const config = {
...
+            test: /\.(sass|scss)$/,
+            include: /\.module\.(sass|scss)$/,
             exclude: storybookDirectory,
+            use: getCSSLoaders('style-loader', {
+                loader: 'css-loader',
+                options: {
+                    sourceMap: isDevelopment,
+                    localsConvention: 'camelCase',
+                    modules: {
+                        localIdentName: isDevelopment ? '[name]__[local]' : '[hash:base64]',
```
Comment (felixfbecker) [inline client/storybook/src/main.ts:138]: Can we include the class name in production too so that it's visible on the hosted Storybook in Chromatic? Might be important there even more than in the webapp production build, as it's a playground for devs where the dev tools are used frequently.

File: client/browser/config/webpack/base.config.ts [hunk 10]
```diff
@@ -69,23 +85,22 @@ export const config: webpack.Configuration = {
...
+            {
+                test: /\.(css|sass|scss)$/,
+                include: /\.module\.(sass|scss)$/,
+                use: getCSSLoaders({
+                    loader: 'css-loader',
+                    options: {
+                        sourceMap: false,
+                        localsConvention: 'camelCase',
+                        modules: {
+                            localIdentName: '[name]__[local]',
```
Comment (felixfbecker) [inline client/browser/config/webpack/base.config.ts:100]: Is there a thought behind not having a hash included here in the browser extension, but to include a hash in the webapp bundle?

**Review by felixfbecker — APPROVED** [review]
LGTM otherwise

File: client/web/webpack.config.js (hunk 7 above)
Comment (felixfbecker) [reply client/web/webpack.config.js:58]: We do need to process SCSS files in `node_modules` for the reason @valerybugakov mentioned, and that shouldn't affect performance as imports are very explicit. If we didn't Bootstrap couldn't respect our Sass variables and we couldn't use the mixins :)

(we do the same with Babel, btw, because otherwise we could not guarantee that imported files from `node_modules` are targetting/supporting the browsers we target/support through `browserslist`/`@babel/preset-env`, and we've had bugs from that)

However this setting is not for that purpose, what it does is allow imports like
```scss
@import 'bootstrap/scss/foo';
```
equivalent to how it works in TS/JS from `node_modules`, as opposed to needing to write out
```scss
@import '../../../node_modules/bootstrap/scss/foo';
```

So this doesn't have any performance consideration.

File: client/browser/config/webpack/base.config.ts (hunk 10 above)
Comment (valerybugakov) [reply client/browser/config/webpack/base.config.ts:100]: Nope, the configuration should be identical. I will update it to match the web-app config.

File: client/storybook/src/main.ts (hunk 9 above)
Comment (valerybugakov) [reply client/storybook/src/main.ts]: Right, ideally, we need one central place for configuration.

**Comment by pdubroy** [comment]
@felixfbecker No, I understood that it was about code intel for the new generated types. But you have a good point about "an expectation or 'standard' for all code added". Maybe this falls into that category, I'm not sure.

In general I just want to encourage keeping PRs small and limited in scope. We shouldn't avoid merging something that makes things better, just because it doesn't make it "better enough". There's obviously a fine line though :-)

**Comment by felixfbecker** [comment]
I totally agree with that.
